File: androidApp/src/main/java/com/cistus/hunter/android/scenes/ToS.kt
package com.cistus.hunter.android.scenes

import android.content.Context
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import com.cistus.hunter.Directory
import com.cistus.hunter.LocalDataConfig
import com.cistus.hunter.TextFile
import com.cistus.hunter.android.Primary2
import com.cistus.hunter.android.R
import com.cistus.hunter.android.UIDraw

class ToS(context: Context) {

    init {
        val tos = LocalDataConfig.ToS(context)
        val tosDirectoryPath = tos.directoryPath
        val tosDirectory = Directory(tosDirectoryPath)
        val tosDataFilePath = tos.dateFilePath
        val tosDataFile = TextFile(tosDataFilePath)
        if (!tosDirectory.isExists())
            tosDirectory.create()
        tosDataFile.write(context.getString(R.string.ToS_Date))
    }

    @Composable
    fun Draw(padding: Float) {
        UIDraw.CenterColumn {
            UIDraw.CenterColumn(fillStyle = UIDraw.FILLSTYLE_MAXWIDTH, modifier = Modifier.height(UIDraw.toDp(padding))) {
                UIDraw.DrawText("利用規約", color = Color.Primary2)
            }
            Box(modifier = Modifier
                .verticalScroll(rememberScrollState()).background(MaterialTheme.colorScheme.background)) {
                ProgressLines(stringResource(R.string.ToS))
            }
        }
    }
    @Composable
    private fun ProgressLines(contents: String) {
        UIDraw.CustomColumn(style = "TopStart", spacing = 3.dp) {
            for (lines in contents.lines()) {
                if (lines.indexOfFirst { it == '*' } >= 0) {
                    val first = lines.indexOfFirst { it == '*' } + 1
                    val last = lines.indexOfLast { it == '*' }
                    val display = lines.substring(first..<last)
                    UIDraw.DrawText(display, color = Color.Primary2, fontSize = 24f,
                        style = "Bold")
                }
                else if (lines.indexOfFirst { it == '|' } >= 0) {
                    val first = lines.indexOfFirst { it == '|' } + 1
                    val last = lines.indexOfLast { it == '|' }
                    val item = lines.substring(first..<last)
                    val content = lines.substring(last + 1)
                    if (item.isEmpty())
                        UIDraw.DrawText("　・${content}", color = Color.Primary2)
                    else
                        UIDraw.DrawText("　${item}. $content", color = Color.Primary2,
                            style = "Bold")
                }
                else
                    UIDraw.DrawText(lines, color = Color.Primary2)
            }
        }
    }

    companion object {
        @Composable
        fun checkUpdate(): Boolean {
            val context = LocalContext.current
            val tos = LocalDataConfig.ToS(context)
            val tosDateFilePath = tos.dateFilePath
            val tosDateFile = TextFile(tosDateFilePath)
            tosDateFile.read()?.let { date ->
                return date != stringResource(R.string.ToS_Date)
            }
            return true
        }
    }
}

File: iosApp/iosApp/Scenes/ToS.swift
import SwiftUI
import shared

struct ToS: View {
    private let close: Binding<Bool>?
    
    init(_ close: Binding<Bool>? = nil) {
        self.close = close
        let tos = LocalDataConfig.ToS()
        let tosDirectoryPath = tos.directoryPath
        let tosDirectory = Directory(path: tosDirectoryPath)
        let tosDateFilePath = tos.dateFilePath
        let tosDateFile = TextFile(path: tosDateFilePath)
        if (!tosDirectory.isExists()) {
            tosDirectory.create()
        }
        tosDateFile.write(contents: String(localized: "Date", table: "ToS"))
    }
    
    var body: some View {
        NavigationStack {
            ScrollView {
                progressLines(contents: String(localized: "ToS", table: "ToS"))
            }
            .toolbar {
                if let close = close {
                    ToolbarItem(placement: .topBarLeading) {
                        Button("閉じる") {
                            close.wrappedValue = false
                        }
                    }
                }
            }
            .navigationTitle("利用規約")
            .navigationBarTitleDisplayMode(.large)
        }
    }
    
    private func progressLines(contents: String) -> some View {
        let lines = contents.split(whereSeparator: \.isNewline)
        return VStack(spacing: 3) {
            ForEach(lines, id: \.self) { line in
                if let first = line.firstIndex(of: "*") {
                    let display = line[line.index(first, offsetBy: 1)..<line.lastIndex(of: "*")!]
                    UIDraw.text(String(display), font: .title2, style: "Bold")
                        .padding(.top, 10)
                        .padding(.leading, 10)
                }
                else if let first = line.firstIndex(of: "|") {
                    let last = line.lastIndex(of: "|")!
                    let item = line[line.index(first, offsetBy: 1)..<line.lastIndex(of: "|")!]
                    let content = line[line.index(last, offsetBy: 1)..<line.endIndex]
                    if (item.count == 0) {
                        UIDraw.text("　・\(content)")
                    }
                    else {
                        UIDraw.text("　\(item). \(content)", style: "Bold")
                    }
                }
                else {
                    UIDraw.text("　\(line)")
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)
        }
    }
    
    static func checkUpdate() -> Bool {
        let tos = LocalDataConfig.ToS()
        let tosDateFilePath = tos.dateFilePath
        let tosDateFile = TextFile(path: tosDateFilePath)
        if let date = tosDateFile.read() {
            if (date != String(localized: "Date", table: "ToS")) {
                return true
            }
        }
        else {
            return true
        }
        return false
    }
}

#Preview {
    ToS()
}
